### PR TITLE
Switch to Concurrent State Machine. Reorganize/rename.

### PR DESCRIPTION
**Pull request by rpiaggio**
Overhaul of the Apollo client to use a concurrent FSM for its internal state, including connection and subscriptions.

There are few edge cases that still need consideration in a future PR.

Also things have been reorganized and renamed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,8 +7,8 @@ Experimental GraphQL client.
 ### 1) Create a client
 
 Either:
-  * A `GraphQLClient[F[_], S]` (supporting queries and mutations), or
-  * A `GraphQLStreamingClient[F[_], S]` (supporting queries, mutations and subscriptions).
+  * A `TransactionalClient[F[_], S]` (supporting queries and mutations), or
```
Comment (cquiroz) [inline README.md:10]: Is there a benefit of using this over `Streaming`? If e.g. you don't plan to streambut maybe want to keep the option open?

File: core/shared/src/main/scala/clue/ApolloClient.scala [hunk 2]
```diff
@@ -1,119 +1,469 @@
...
 
-object ApolloClient {
-  type ConnectionDef[F[_], CP] = TryableDeferred[F, PersistentConnection[F, CP]]
-  // Connection =
-  //   - None: Not yet connected or disconnected (either by client or by server and client gave up retrying).
-  //   - Some(incomplete Deferred): Connection requested but not established.
-  //   - Some(complete Deferred): Connection established.
-  type Connection[F[_], CP]    = Option[ConnectionDef[F, CP]]
+// Client internal state for the FSM.
+protected sealed abstract class State[+F[_], +CP](val status: PersistentClientStatus)
```
Comment (cquiroz) [inline core/shared/src/main/scala/clue/ApolloClient.scala:38]: Nice

File: core/shared/src/main/scala/clue/ApolloClient.scala [hunk 3]
```diff
@@ -1,119 +1,469 @@
...
-  private val getConnectionDef: F[ApolloClient.ConnectionDef[F, CP]] =
-    firstInitInvoked.get >>
-      connectionRef.get.flatMap(_ match {
-        case None                => F.raiseError(new Exception("Client is disconnected"))
-        case Some(connectionDef) => F.pure(connectionDef)
-      })
+    Latch[F].flatMap { newLatch =>
+      stateModify {
+        case Disconnected                              => Connecting(newLatch) -> doConnect(newLatch)
+        case s @ Connecting(latch)                     => s                    -> (warn >> latch.get.rethrow)
```
Comment (cquiroz) [inline core/shared/src/main/scala/clue/ApolloClient.scala:106]: Any reason to use `>>` over `*>`?

File: README.md (hunk 1 above)
Comment (rpiaggio) [reply README.md:10]: Not sure I understood. The renames were:
* `GraphQLClient` (which was a bit broad) -> `TransactionalClient` (meaning client that supports one-shots but not streaming, alternate name suggestions are welcome).
* `GraphQLStreamingClient` -> `StreamingClient`.

In general I tried removing the `GraphQL` prefix as much as possible.

File: core/shared/src/main/scala/clue/package.scala [hunk 4]
```diff
@@ -1,13 +1,65 @@
...
+
+  type WebSocketBackend[F[_]]    = PersistentBackend[F, WebSocketCloseParams, WebSocketCloseEvent]
+  type WebSocketConnection[F[_]] = PersistentConnection[F, WebSocketCloseParams]
+
+  type ApolloWebSocketClient[F[_], S] =
+    ApolloClient[F, S, WebSocketCloseParams, WebSocketCloseEvent]
+
+  protected[clue] type Latch[F[_]] = Deferred[F, Either[Throwable, Unit]]
+
+  final implicit class StringOps(val str: String) extends AnyVal {
```
Comment (cquiroz) [inline core/shared/src/main/scala/clue/package.scala:29]: Maybe this is a general utility we could use in other places

**Review by cquiroz — APPROVED** [review]
Looking great, the concept of FSM applies very well to manage connections. I haven't checked every detail in the code but in general it seems great

File: core/shared/src/main/scala/clue/ApolloClient.scala (hunk 3 above)
Comment (rpiaggio) [reply core/shared/src/main/scala/clue/ApolloClient.scala:106]: In general, I prefer `>>` when things should be run in sequence, making sure we are in the `Monad` world and not in an `Applicative`. Also, the parameter to `>>` is lazily evaluated, while `>*`'s is eager, which may be an issue when there's recursion involved.

Fabio explains it better than me here: https://gitter.im/typelevel/cats?at=5bbf6d49bbdc0b250523988c

File: README.md (hunk 1 above)
Comment (cquiroz) [reply README.md:10]: I meant of using `TransactionalClient` over `StreamingClient`. does it make sense to have a single one?

Maybe it is about allowing implementations that don't support streaming?

File: core/shared/src/main/scala/clue/package.scala (hunk 4 above)
Comment (rpiaggio) [reply core/shared/src/main/scala/clue/package.scala:29]: Sure, if other people like this syntax I'll be happy to move it (or copy it? I'm not sure I want to introduce non-general dependencies).

File: README.md (hunk 1 above)
Comment (rpiaggio) [reply README.md:10]: Exactly. If you want to do one-shots over HTTP, that would be the one to use.

**Pull request merged**